src/intrin_portable.h

```diff
@@ -151,6 +151,211 @@ FORCE_INLINE void rx_set_rounding_mode(uint32_t mode) {
+
+/* Splat 32-bit int to 4 32-bit ints */
+static inline __m128i vec_splat4sw (int32_t scalar)
+{ return (__m128i) vec_splats (scalar); }
+
+/* Splat 64-bit long long to 2 64-bit long longs */
+static inline __m128i vec_splat2sd (int64_t scalar)
+{ return (__m128i) vec_splats (scalar); }
+
+FORCE_INLINE rx_vec_f128 rx_load_vec_f128(const double* pd) {
+	rx_vec_f128 x;
+  uint64_t* xiu64 = (uint64_t*)&x;
+	xiu64[0] = load64(pd + 0);
+	xiu64[1] = load64(pd + 1);
+	return x;
+}
+
+FORCE_INLINE void rx_store_vec_f128(double* mem_addr, rx_vec_f128 a) {
+  uint64_t* aiu64 = (uint64_t*)&a;
+  store64(mem_addr + 0, aiu64[0]);
+	store64(mem_addr + 1, aiu64[1]);
+}
+
+FORCE_INLINE rx_vec_f128 rx_swap_vec_f128(rx_vec_f128 a) {
+	return vec_perm(a,a,(__m128i){8,9,10,11,12,13,14,15,0,1,2,3,4,5,6,7});
+}
+
+FORCE_INLINE rx_vec_f128 rx_add_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_add(a,b);
+}
+
+FORCE_INLINE rx_vec_f128 rx_sub_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_sub(a,b);
+}
+
+FORCE_INLINE rx_vec_f128 rx_mul_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_mul(a,b);
+}
+
+FORCE_INLINE rx_vec_f128 rx_div_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_div(a,b);
+}
+
+FORCE_INLINE rx_vec_f128 rx_sqrt_vec_f128(rx_vec_f128 a) {
+	return vec_sqrt(a);
+}
+
+FORCE_INLINE rx_vec_i128 rx_set1_long_vec_i128(uint64_t a) {
+	return vec_splat2sd(a);
+}
+
+FORCE_INLINE rx_vec_f128 rx_vec_i128_vec_f128(rx_vec_i128 a) {
+	return (rx_vec_f128)a;
+}
+
+FORCE_INLINE rx_vec_f128 rx_set_vec_f128(uint64_t x1, uint64_t x0) {
+	return (rx_vec_f128)(__m128ll){x0,x1};
+}
+
+FORCE_INLINE rx_vec_f128 rx_set1_vec_f128(uint64_t x) {
+	return (rx_vec_f128)vec_splat2sd(x);
+}
+
+
+FORCE_INLINE rx_vec_f128 rx_xor_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_xor(a,b);
+}
+
+FORCE_INLINE rx_vec_f128 rx_and_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_and(a,b);
+}
+
+FORCE_INLINE rx_vec_f128 rx_or_vec_f128(rx_vec_f128 a, rx_vec_f128 b) {
+	return vec_or(a,b);
+}
+#if defined(__CRYPTO__)
+
+FORCE_INLINE rx_vec_i128 vrev(rx_vec_i128 v){
+#if defined(NATIVE_LITTLE_ENDIAN)
+return vec_perm(v,__m128i{0},__m128i{15,14,13,12,11,10,9,8,7,6,5,4,3,2,1,0});
+#else
+return v;
+#endif
+}
+
+FORCE_INLINE rx_vec_i128 rx_aesenc_vec_i128(rx_vec_i128 v, rx_vec_i128 rkey) {
+  return vrev(__builtin_crypto_vcipher(vrev(v),vrev(rkey)));
+}
+
+FORCE_INLINE rx_vec_i128 rx_aesdec_vec_i128(rx_vec_i128 v, rx_vec_i128 rkey) {
+  rx_vec_i128 out = vrev(__builtin_crypto_vncipher(vrev(v),(__m128i){0}));
+  return vec_xor(out,rkey); 
+}
+#else
+static const char* platformError = "Platform doesn't support hardware AES";
+
+FORCE_INLINE rx_vec_i128 rx_aesenc_vec_i128(rx_vec_i128 v, rx_vec_i128 rkey) {
+	throw std::runtime_error(platformError);
+}
+
+FORCE_INLINE rx_vec_i128 rx_aesdec_vec_i128(rx_vec_i128 v, rx_vec_i128 rkey) {
+	throw std::runtime_error(platformError);
+}
+#endif
+
+
+FORCE_INLINE int rx_vec_i128_x(rx_vec_i128 a) {
+	return ((int*)&a)[0];
```

Comment:
It's better to do this using a `union` to prevent aliasing warnings from the compiler.

```c++
union vec_i32 {
    rx_vec_i128 vec;
    int i32[4];
};
```